### ISSUE TITLE
Dangerous behaviour of timestring

### ISSUE THREAD

**Issue by vbohata** [elastic/curator#1271]
According to docs the timestring finds pattern anywhere in the index name. Nowadays it is best practice for log management indexes to include beats version in index names. We also include our doc version in index name and use not only beats as agents. So index names can be:
- index-myversion-50-filebeat-6.3.0-2018.05
- index-myversion-10-customagent-12.5-2018.05

and easily become:
- index-myversion-10-customagent-2019.11-2018.05

As it is more and more favourite to do big number steps (or use year versioning) in app versions.

I propose to add config option to allow timestring position. In most cases the date will be the last part of index name so by default it could also be fixed to "DATE$" pattern.

I think it is a matter of time this behaviour will cause a big troubles.

**Comment by untergeek** [comment]
You are correct, that these big strings of dot-separated numbers are problematic.  However, these indices are _always_ created by the `*beat`, which means they are created according to UTC time, which means that you will be far better served by not using the `timestring` at all, and use `source: creation_date` instead of `source: name` for your `age` filters.

`source: name` is throwback to before the era of `*beats`, when Logstash was the only game in town, and created `logstash-YYYY.MM.dd` indices.  This pattern still exists for when `source: creation_date` is insufficient.  On top of this, we encourage the use of rollover indices instead of time-based, which further removes you from indices with a timestamp in the name.  In these cases `source: creation_date` or better yet, [`source: field_stats`](https://www.elastic.co/guide/en/elasticsearch/client/curator/5.5/filtertype_age.html#_literal_field_stats_literal_based_ages) for time range selection is a preferable choice.

Curator also has the [`period` filter](https://www.elastic.co/guide/en/elasticsearch/client/curator/current/filtertype_period.html), which would accommodate your month-range example.

While I concur that using `source: name` can be problematic with complex index names, these other options are available to you which are superior, and obviate the need to add complicated regex anchoring to the Curator syntax.

**Comment by vbohata** [comment]
For some indexes I already use period filter. Our agent (beats/nxlog/...) indexes are always created by logstash as we have a big environment and use message broker etc ... not important for this case.

In the past I used creation_date but had to switch to timestring.
1. It is not usable as we often index old data (older then month for example) to appropriate indexes. So even if index creation date is today, it actually contains old data. Also sometimes we need to reindex the data and each reindex create a new index with current creation date even for very old data. So the only solution is to use time based index naming and timestring filtering.
2. Filtering by creation_date is impossible so if not using timestring I have to use field_stats which is far slower (hours vs. seconds for thousands of indexes). Not possible to use - is slow and generates significant load.

**Comment by untergeek** [comment]
If it takes hours for even a few thousand individual field_stats queries to come back, that's a problem.  It should not be taking that long.

See https://github.com/elastic/curator/blob/v5.5.4/curator/indexlist.py#L307-L329
```
        for l in index_lists:
            for index in l:
                body = {
                    'aggs' : {
                        'min' : { 'min' : { 'field' : field } },
                        'max' : { 'max' : { 'field' : field } }
                    }
                }
                response = self.client.search(index=index, size=0, body=body)
                self.loggit.debug('RESPONSE: {0}'.format(response))
                if response:
                    try:
                        r = response['aggregations']
                        self.loggit.debug('r: {0}'.format(r))
                        s = self.index_info[index]['age']
                        s['min_value'] = utils.fix_epoch(r['min']['value'])
                        s['max_value'] = utils.fix_epoch(r['max']['value'])
                        self.loggit.debug('s: {0}'.format(s))
                    except KeyError:
                        raise exceptions.ActionError(
                            'Field "{0}" not found in index '
                            '"{1}"'.format(field, index)
                            )
```
It loops over all indices in the current working list and does a small aggregation query to find the min and max.  This should not take that long.  You're telling me that it's taking hours to execute this across thousands of indices? Why are you not pre-filtering these indices with pattern filters to reduce the overhead?

I, personally, will not add this requested feature for the following reasons:
* This is a limited scope use case that few users will need, irrespective of how pressing your need may be
* Curator is about to be superseded by Index Lifecycle Management, though that will require at least a Basic (free) license.  As such, Curator is in more of a maintenance mode, with patches, and limited new functionality being added. Curator will continue to exist for edge cases that ILM does not or will not cover.

If someone else wants to fork the code and add this feature, and add tests, and it passes code review, I will likely merge it.

**Comment by untergeek** [comment]
I can't comment on release dates, unfortunately.

**Comment by untergeek** [comment]
> Everyone who is using timestring needs it. The current behaviour is not a problem until it will become ... This is the type of issue when everything is OK, everything is OK, ... WTF! why my indexes disappears, where are my data ....

You're the first user in my 5 years at Elastic who has tried to make this sound like every user will be affected.  A few have been affected, but all of them have been okay with switching to `creation_date` or `field_stats`.  You would be the first user who—for reasons already described—cannot.  So why not switch the date separator character so it's always uniform, but differ significantly from the other index name bits? Why not use multiple characters? `%Y.%m.%d` isn't mandatory. It could just as easily be `%Y_._%m_._%d` or any other permutation you can imagine.  There are many easier ways to solve this that don't require a huge refactoring of Curator's code.

**Comment by vbohata** [comment]
Maybe it will be easier to solve it. In curator/utils.py get_date_regex seems to allow arbitrary char, also TimestringSearch should be handle it so I should be able to set timestring to "%Y.%m.%d$". Am I right or is there anything I do not see?

**Comment by untergeek** [comment]
It would only be able to anchor the beginning or end of the line, but I don't see why it wouldn't work.
